src/connectors/impls/gpubsub/gsub.rs

```diff
@@ -0,0 +1,279 @@
+use crate::connectors::google::AuthInterceptor;
+use crate::connectors::prelude::*;
+use async_std::channel::{Receiver, Sender};
+use async_std::prelude::FutureExt;
+use async_std::sync::Mutex;
+use dashmap::DashMap;
+use googapis::google::pubsub::v1::subscriber_client::SubscriberClient;
+use googapis::google::pubsub::v1::{
+    AcknowledgeRequest, PubsubMessage, PullRequest, ReceivedMessage,
+};
+use gouth::Token;
+use serde::Deserialize;
+use std::fmt::Debug;
+use std::sync::Arc;
+use std::time::Duration;
+use tonic::codegen::InterceptedService;
+use tonic::transport::{Certificate, Channel, ClientTlsConfig};
+use tonic::Status;
+use tremor_pipeline::ConfigImpl;
+
+#[derive(Deserialize, Clone)]
+struct Config {
+    pub connect_timeout: u64,
+    pub subscription_id: String,
+}
+impl ConfigImpl for Config {}
+
+#[derive(Debug, Default)]
+pub(crate) struct Builder {}
+
+#[async_trait::async_trait]
+impl ConnectorBuilder for Builder {
+    fn connector_type(&self) -> ConnectorType {
+        "gsub".into()
+    }
+
+    async fn build(&self, alias: &str, raw_config: &ConnectorConfig) -> Result<Box<dyn Connector>> {
+        if let Some(raw) = &raw_config.config {
+            let config = Config::new(raw)?;
+            Ok(Box::new(GSub { config }))
+        } else {
+            Err(ErrorKind::MissingConfiguration(alias.to_string()).into())
+        }
+    }
+}
+
+struct GSub {
+    config: Config,
+}
+
+type PubSubClient = SubscriberClient<InterceptedService<Channel, AuthInterceptor>>;
+type AsyncTaskMessage = Result<(u64, Vec<u8>)>;
+
+struct GSubSource {
+    config: Config,
+    client: Option<Arc<Mutex<PubSubClient>>>,
+    receiver: Option<Receiver<AsyncTaskMessage>>,
+    ack_ids: Arc<DashMap<u64, String>>,
+}
+
+impl GSubSource {
+    pub fn new(config: Config) -> Self {
+        GSubSource {
+            config,
+            client: None,
+            receiver: None,
+            ack_ids: Arc::new(DashMap::new()),
+        }
+    }
+}
+
+async fn consumer_task(
+    client: Arc<Mutex<PubSubClient>>,
+    sender: Sender<AsyncTaskMessage>,
+    ack_ids: Arc<DashMap<u64, String>>,
+    subscription_id: String,
+    request_timeout: Duration,
+) {
+    let mut ack_counter = 0;
+
+    loop {
+        let response_with_potential_timeout = client
+            .lock()
```

Comment:
i am afraid that this lock will block the `Source` from handling `ack/fail`. I think you can simply clone the `client` and pass a clone into this task, no? This would avoid the `Arc<Mutex<...>>` and leave the dashmap as the only synchronization point.

Reply:
Turns out that the client can't be cloned, but it seems like a sensible workaround would be to create two clients using a cloned channel (according to tonic's docs, this will still use one connection handled in a background task)